exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DynamicRootSchema.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.calcite.schema.impl.AbstractSchema;
+import org.apache.calcite.util.BuiltInMethod;
+import org.apache.calcite.util.Compatible;
+import org.apache.drill.common.exceptions.ExecutionSetupException;
+import org.apache.drill.exec.store.SchemaConfig;
+import org.apache.drill.exec.store.StoragePlugin;
+import org.apache.drill.exec.store.StoragePluginRegistry;
+import org.apache.drill.exec.store.SubSchemaWrapper;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Set;
+
+public class DynamicRootSchema extends DynamicSchema
+    implements CalciteRootSchema {
+
+  /** Creates a root schema. */
+  DynamicRootSchema(StoragePluginRegistry storages, SchemaConfig schemaConfig) {
+    super(null, new RootSchema(), "");
+    this.schemaConfig = schemaConfig;
+    this.storages = storages;
+  }
+
+  @Override
+  public CalciteSchema getSubSchema(String schemaName, boolean caseSensitive) {
+    CalciteSchema retSchema = getSubSchemaMap().get(schemaName);
+
+    if (retSchema == null) {
+      loadSchemaFactory(schemaName, caseSensitive);
+    }
+
+    retSchema = getSubSchemaMap().get(schemaName);
+    return retSchema;
+  }
+
+  @Override
+  public NavigableSet<String> getTableNames() {
+    Set<String> pluginNames = Sets.newHashSet();
+    for (Map.Entry<String, StoragePlugin> storageEntry : getSchemaFactories()) {
+      pluginNames.add(storageEntry.getKey());
+    }
+    return Compatible.INSTANCE.navigableSet(
+        ImmutableSortedSet.copyOf(
+            Sets.union(pluginNames, getSubSchemaMap().keySet())));
+  }
+
+  /**
+   * load schema factory(storage plugin) for schemaName
+   * @param schemaName
+   * @param caseSensitive
+   */
+  public void loadSchemaFactory(String schemaName, boolean caseSensitive) {
+    try {
+      SchemaPlus thisPlus = this.plus();
+      StoragePlugin plugin = getSchemaFactories().getPlugin(schemaName);
+      if (plugin != null) {
+        plugin.registerSchemas(schemaConfig, thisPlus);
+      }
+      else {
+        //this schema name could be `dfs.tmp`, a 2nd level schema under 'dfs'
+        String[] paths = schemaName.split("\\.");
+        if (paths.length == 2) {
+          plugin = getSchemaFactories().getPlugin(paths[0]);
+          if (plugin != null) {
+            plugin.registerSchemas(schemaConfig, thisPlus);
+          }
+
+          final CalciteSchema secondLevelSchema = getSubSchemaMap().get(paths[0]).getSubSchema(paths[1], caseSensitive);
+          if (secondLevelSchema != null) {
+            SchemaPlus secondlevel = secondLevelSchema.plus();
+            org.apache.drill.exec.store.AbstractSchema drillSchema =
+                secondlevel.unwrap(org.apache.drill.exec.store.AbstractSchema.class);
+            SubSchemaWrapper wrapper = new SubSchemaWrapper(drillSchema);
+            thisPlus.add(wrapper.getName(), wrapper);
+          }
+        }
+      }
+    } catch(ExecutionSetupException | IOException ex) {
```

Comment:
Really? Just ignore errors? Maybe add a comment inside the catch block to explain why.